### PR TITLE
RQT Parameters for Ros2

### PR DESCRIPTION
**Pull request by gonzodepedro**

### CODE REVIEW

**Comment by cottsay** [comment]
Let's create a `dashing` branch for these changes. I'll create the branch from the current `master` - if you could re-create this PR against that branch so the changes can be reviewed, that would be great.

**Comment by gonzodepedro** [comment]
Done

File: package.xml [hunk 1]
```diff
@@ -1,6 +1,6 @@
-<package>
+<package format="3">
 	<name>rqt_reconfigure</name>
-	<version>0.4.10</version>
+	<version>1.4.10</version>
```
Comment (cottsay) [inline package.xml]: What is the logic behind this version bump?

File: package.xml [hunk 2]
```diff
@@ -23,17 +23,18 @@
 	<author>Isaac Saito</author>
 	<author>Ze'ev Klapow</author>
 
-	<buildtool_depend>catkin</buildtool_depend>
+	<buildtool_depend>ament_cmake</buildtool_depend>
```
Comment (cottsay) [inline package.xml]: I don't see anything fancy happening in the `CMakeLists.txt` - is there a reason not to just make this an `ament_python` package?

File: package.xml (hunk 1 above)
Comment (gonzodepedro) [reply package.xml]: Fixed

File: package.xml (hunk 2 above)
Comment (gonzodepedro) [reply package.xml]: I agree. Changed to 'ament_python"

**Comment by dirk-thomas** [comment]
The current patch mixes a lot of unrelated changes. E.g. whitespace / formatting changes should be done separately - they could land in master - and then this PR will focus on changes related to ROS 2. The same goes for other non-ROS 2 changes like changing the permission flags of the test files.

File: package.xml (hunk 2 above)
Comment (cottsay) [reply package.xml]: This didn't happen. An `ament_python` package doesn't include a `CMakeLists.txt`.

We'll also need a more robust `setup.py` like this:
```python
from setuptools import setup

package_name = 'rqt_reconfigure'
setup(
    name=package_name,
    version='1.0.1',
    package_dir={'': 'src'},
    packages=[package_name],
    data_files=[
        ('share/ament_index/resource_index/packages',
            ['resource/' + package_name]),
        ('share/' + package_name, ['package.xml', 'plugin.xml']),
        ('share/' + package_name + '/resource', [
            'resource/editor_bool.ui',
            'resource/editor_enum.ui',
            'resource/editor_number.ui',
            'resource/editor_string.ui',
            'resource/node_selector.ui',
            'resource/paramedit_pane.ui',
            'resource/param_main.ui',
            'resource/param_timeline.ui',
            'resource/singlenode_parameditor.ui',
            'resource/text_filter_widget.ui',
        ]),
    ],
    zip_safe=True,
    author='Isaac Saito, Ze\'ev Klapow',
    maintainer='Scott K Logan',
    maintainer_email='logans@cottsay.net',
    keywords=['ROS'],
    classifiers=[
        'Intended Audience :: Developers',
        'License :: OSI Approved :: BSD License',
        'Programming Language :: Python',
        'Topic :: Software Development',
    ],
    description=(
        'This rqt plugin succeeds former dynamic_reconfigure\'s GUI ' +
        '(reconfigure_gui), and provides the way to view and edit the parameters ' +
        'that are accessible via dynamic_reconfigure.'
    ),
    license='BSD',
    entry_points={
        'console_scripts': [
            'rqt_reconfigure = rqt_reconfigure.main:main',
        ],
    },
)
```

File: package.xml [hunk 3]
```diff
@@ -23,17 +23,18 @@
...
+
+  <exec_depend version_gte="0.2.19">python_qt_binding</exec_depend>
+	<exec_depend>python3-catkin-pkg-modules</exec_depend>
+	<exec_depend>qt_gui_py_common</exec_depend>
+	<exec_depend>rqt_gui</exec_depend>
+	<exec_depend>rqt_gui_py</exec_depend>
+  <exec_depend>rqt_py_console</exec_depend>
```
Comment (cottsay) [inline package.xml]: Mixed tabs and spaces here and on the first `<exec_depend>`

File: package.xml [hunk 4]
```diff
@@ -23,17 +23,18 @@
...
 
-	<run_depend>dynamic_reconfigure</run_depend>
-  <run_depend version_gte="0.2.19">python_qt_binding</run_depend>
-	<run_depend>rospy</run_depend>
-	<run_depend>rqt_console</run_depend>
-	<run_depend>rqt_gui</run_depend>
-	<run_depend>rqt_gui_py</run_depend>
-	<run_depend>rqt_py_common</run_depend>
 	<export>
-		<architecture_independent/>
```
Comment (cottsay) [inline package.xml]: This package is still architecture independent.

File: test/test_parameters_service.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+
+import unittest
+import rclpy
+from rqt_reconfigure.parameters_services import ParametersServices
```
Comment (cottsay) [inline test/test_parameters_service.py]: I can't find any mention of this in `src`.

Comment (gonzodepedro) [reply test/test_parameters_service.py]: Removed

File: package.xml (hunk 4 above)
Comment (gonzodepedro) [reply package.xml]: Got it

File: package.xml (hunk 2 above)
Comment (gonzodepedro) [reply package.xml]: Done

File: src/rqt_reconfigure/param_editors.py [hunk 6]
```diff
@@ -32,37 +32,22 @@
...
-                       'editor_bool.ui')
-ui_str = os.path.join(rp.get_path('rqt_reconfigure'), 'resource',
-                      'editor_string.ui')
-ui_num = os.path.join(rp.get_path('rqt_reconfigure'), 'resource',
-                      'editor_number.ui')
-ui_int = ui_num
-ui_enum = os.path.join(rp.get_path('rqt_reconfigure'), 'resource',
-                       'editor_enum.ui')
-
+_, package_path = get_resource('packages', 'rqt_graph')
```
Comment (cottsay) [inline src/rqt_reconfigure/param_editors.py]: This is causing errors every time I try to connect to a node.
```suggestion
_, package_path = get_resource('packages', 'rqt_reconfigure')
```

File: src/rqt_reconfigure/param_editors.py [hunk 7]
```diff
@@ -185,32 +173,41 @@ def __init__(self, updater, config):
...
+        super(IntegerEditor, self).__init__(*args, **kwargs)
+        ui_int = os.path.join(
+            package_path, 'share', 'rqt_reconfigure', 'resource', 'editor_number.ui'
+        )
         loadUi(ui_int, self)
 
         # Set ranges
-        self._min = int(config['min'])
-        self._max = int(config['max'])
+        self._min = int(self.descriptor.integer_range[0].from_value)
```
Comment (cottsay) [inline src/rqt_reconfigure/param_editors.py]: This line generates an error whenever there is an `IntegerEditor` present:
```
  File "src/rqt_reconfigure/src/rqt_reconfigure/param_editors.py", line 207, in __init__
    self._min = int(self.descriptor.integer_range[0].from_value)
IndexError: list index out of range
```

File: src/rqt_reconfigure/param_editors.py [hunk 8]
```diff
@@ -269,35 +273,48 @@ def _update_gui(self, value):
...
-        super(DoubleEditor, self).__init__(updater, config)
+    def __init__(self, *args, **kwargs):
+        super(DoubleEditor, self).__init__(*args, **kwargs)
+        ui_num = os.path.join(
+            package_path, 'share', 'rqt_reconfigure', 'resource', 'editor_number.ui'
+        )
         loadUi(ui_num, self)
 
+        #config step
+        self._step = float(self.descriptor.floating_point_range[0].step)
```
Comment (cottsay) [inline src/rqt_reconfigure/param_editors.py]: Similar to the `IntegerEditor` issue, a similar error when using `DoubleEditor`

```
  File "/home/cottsay/rqt_ws/build/rqt_reconfigure/src/rqt_reconfigure/param_editors.py", line 299, in __init__
    self._step = float(self.descriptor.floating_point_range[0].step)
IndexError: list index out of range
```

File: src/rqt_reconfigure/node_selector_widget.py [hunk 9]
```diff
@@ -38,47 +38,51 @@
...
 
-        ui_file = os.path.join(rospack.get_path('rqt_reconfigure'), 'resource',
+        self.stretch = None
+        _, package_path = get_resource('packages', 'rqt_reconfigure')
+        ui_file = os.path.join(package_path, 'share', 'rqt_reconfigure', 'resource',
                                'node_selector.ui')
         loadUi(ui_file, self)
 
-        # List of the available nodes. Since the list should be updated over
+        # List of the availa98ble nodes. Since the list should be updated over
```
Comment (cottsay) [inline src/rqt_reconfigure/node_selector_widget.py]: Are you using VIM by chance? :wink:

File: plugin.xml [hunk 10]
```diff
@@ -1,5 +1,5 @@
 <library path="src">
-  <class name="Param" type="rqt_reconfigure.param_plugin.ParamPlugin" base_class_type="rqt_gui_py::Plugin">
+  <class name="ParamPlugin" type="rqt_reconfigure.param_plugin.ParamPlugin" base_class_type="rqt_gui_py::Plugin">
```
Comment (cottsay) [inline plugin.xml]: I'm curious about this change - is it to conform to a pattern? Should it be made in ROS 1 as well?

File: src/rqt_reconfigure/filter_children_model.py [hunk 11]
```diff
@@ -119,8 +119,8 @@ def _filter_row_recur(self, src_row, src_parent_qmindex):
 
             # If the index is the terminal treenode, parameters that hit
             # the query are displayed at the root tree.
-            _child_index = curr_qmindex.child(0, 0)
-            if ((not _child_index.isValid()) and
+            child_index = curr_qmindex.child(0, 0)
+            if ((not child_index.isValid()) and
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/filter_children_model.py]: Please either revert unnecessary / unrelated changes like this or apply the same diff on the ROS 1 branch.

File: src/rqt_reconfigure/node_selector_widget.py [hunk 12]
```diff
@@ -328,8 +330,11 @@ def _update_nodetree_pernode(self):
 
                 # Instantiate QStandardItem. Inside, dyn_reconf client will
                 # be generated too.
+
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/node_selector_widget.py]: Please revert unnecessary / unrelated changes like this to keep the diff between the branches minimal.

File: src/rqt_reconfigure/param_api.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
+# Software License Agreement (BSD License)
+#
+# Copyright (c) 2012, Willow Garage, Inc.
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/param_api.py]: If this file is new the copyright should be OSRF.

File: src/rqt_reconfigure/param_api.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+from rclpy.parameter import Parameter
+from rclpy.qos import qos_profile_parameter_events
+
+
+class ParamClient(object):
+
+    def __init__(self, node, remote_node_name, param_change_callback=None):
+        """
+        """
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/param_api.py]: Please either remove the empty docblock or add actual documentation to it.

File: src/rqt_reconfigure/param_client_widget.py [hunk 15]
```diff
@@ -32,90 +32,128 @@
...
+        self._editor_widgets = {}
+
+        self._param_client = create_param_client(
+            context.node, node_name, self._handle_param_event
+        )
+
+        # TODO: .ui file needs to be back into usage in later phase.
+        # ui_file = os.path.join(rp.get_path('rqt_reconfigure'),
+        #                        'resource', 'singlenode_parameditor.ui')
+        # loadUi(ui_file, self)
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/param_client_widget.py]: What about this TODO?

File: src/rqt_reconfigure/param_client_widget.py [hunk 16]
```diff
@@ -32,90 +32,128 @@
...
+        # Again, these UI operation above needs to happen in .ui file.
+        param_names = self._param_client.list_parameters()
+        self.add_editor_widgets(
+            self._param_client.get_parameters(param_names),
+            self._param_client.describe_parameters(param_names)
+        )
+
+        # Labels should not stretch
+        # self.grid.setColumnStretch(1, 1)
+        # self.setLayout(self.grid)
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/param_client_widget.py]: Why is this commented out?

File: src/rqt_reconfigure/param_editors.py [hunk 17]
```diff
@@ -189,62 +175,87 @@ def __init__(self, updater, config):
...
+            self._slider_horizontal.valueChanged.connect(self._slider_changed)
+
+            # Add special menu items
+            self.cmenu.addAction(self.tr('Set to Maximum')
+                                 ).triggered.connect(self._set_to_max)
+            self.cmenu.addAction(self.tr('Set to Minimum')
+                                 ).triggered.connect(self._set_to_min)
+
+            # TODO: Fix that the naming of _paramval_lineEdit instance is not
+            #       consistent among Editor's subclasses.
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/param_editors.py]: What about this TODO?

File: src/rqt_reconfigure/param_widget.py [hunk 18]
```diff
@@ -78,53 +74,51 @@ def __init__(self, context, node=None):
...
-        rp = rospkg.RosPack()
-
         # TODO: .ui file needs to replace the GUI components declaration
         #       below. For unknown reason, referring to another .ui files
         #       from a .ui that is used in this class failed. So for now,
         #       I decided not use .ui in this class.
         #       If someone can tackle this I'd appreciate.
-        _hlayout_top = QHBoxLayout(self)
-        _hlayout_top.setContentsMargins(QMargins(0, 0, 0, 0))
+        hlayout_top = QHBoxLayout(self)
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/param_widget.py]: What is the reason to rename all these variables?

Please either revert unnecessary / unrelated changes like this or apply the same diff on the ROS 1 branch.

File: src/rqt_reconfigure/paramedit_widget.py [hunk 19]
```diff
@@ -149,17 +143,18 @@ def _remove_node(self, node_grn):
...
+        w = self._param_client_widgets.pop(node_grn)
 
         logging.debug('popped={} Len of left clients={}'.format(
-            w, len(self._param_clients)
+            w, len(self._param_client_widgets)
         ))
 
     def _node_disabled(self, node_grn):
         logging.debug('paramedit_w _node_disabled grn={}'.format(node_grn))
+        # self._checkbox.setChecked(value) # TODO(Gonzo) signal doesn't work
```
Comment (dirk-thomas) [inline src/rqt_reconfigure/paramedit_widget.py]: What about this comment / TODO?
